### ISSUE TITLE
Controller: dealing with multiple instances of the same object

### ISSUE THREAD

**Issue by vmicheli** [alfworld/alfworld#16]
Hi,

Another week another issue 😅

In the Alfworld environment, multiple instances of the same object are assigned different IDs. However, I'm wondering how we could do this identification in Alfred and have it be consistent with Alfworld?

It looks like this would be a necessary step if we were to train a Controller module mapping high-level actions in Alfworld to sequences of low-level actions in Alfred. If I'm not mistaken, Alfred currently uses metadata relative to the `location` of the target object but the `discrete action` does not specify which instance the agent should go to.

Do you have any recommendations or pointers to the codebase in order to solve this issue?

**Comment by MohitShridhar** [comment]
@vmicheli in ALFWorld, we use object IDs `<object_name> <obj_id>` to handle multiple objects of the same class, e.g. `apple 1` and `apple 2`. For all ALFWorld tasks, if the high-level goal is to "put an apple on the plate", it doesn't matter which of the two apples you pick, as long as you put _some_ apple on _some_ plate you will succeed at the task. So the IDs here help in keeping track of objects and receptacles but isn't directly relevant to the task. 

However in ALFRED, there are specific referring expressions like "pick up the red apple on the left", and if you want complete the task as quickly as possible, you should pick the exact apple specified in the low-level language instruction. It's a bit nuanced, but ALFWorld problem and ALFRED problem are slightly different.

**Comment by vmicheli** [comment]
Thanks for your reply @MohitShridhar.

I'm still struggling to understand the case where we have multiple instances of a receptacle. Let's say we have `drawer 1 ` and `drawer 2` and the text agent states `go to drawer 1`. Do you agree that the burden then falls on the controller to pick one of the two drawers and to somehow remember that it will be referred as `drawer 1` in the future? Otherwise, if the controller always picked one drawer at random, the text agent would receive different state estimations for the same `go to drawer 1` instruction and would not know which drawers have already been visited.

**Comment by MohitShridhar** [comment]
@vmicheli yes, that's right. Some "object searches" will become exhaustive in the sense that you have to look through every `drawer` to find the `watch` (but at least you are looking inside drawers and not refrigerators).

Initially, we planned on incorporating some spatial-relationships or other attributes to disambiguate multiple receptacles, e.g. the drawer below the microwave, but we didn't get around to it. If you are interested in this, here is a guide: https://github.com/alfworld/alfworld/tree/master/alfworld/data#data--generation (not through though). It's not trivial, but it should be possible to incorporate additional metadata about objects like 6D pose, material info into the TextWorld observations.

**Comment by vmicheli** [comment]
I see, thanks for the explanation and the pointer @MohitShridhar. I'll let you know if I come up with something!
